src/evaluators/SwitchStatement.js

```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
   return Environment.GetValue(realm, exprRef);
 }
 
+function GetValue(input: Value | Completion, defaultValue: Value, realm: Realm): Value | Completion {
+  // extract the internal value
+  let value;
+  if (input instanceof Completion) {
+    value = input.value;
+  } else {
+    value = input;
+  }
+
+  // if the value we got was not empty, let's use it as the new propagating value
+  if (!(value instanceof EmptyValue)) {
+    defaultValue = value;
+  }
+
+  return UpdateEmpty(realm, input, defaultValue);
+}
+
 function AbstractCaseBlockEvaluation(
   cases: Array<BabelNodeSwitchCase>,
-  defaultCaseNum: number,
+  defaultCaseIndex: number,
   input: Value,
   strictCode: boolean,
   env: LexicalEnvironment,
   realm: Realm
 ): Value {
   invariant(realm.useAbstractInterpretation);
+  let ReportNoReferenceError = () => {
+    let msg = "This operation is not yet supported on case blocks that result in a Reference.";
+    realm.reportIntrospectionError(msg);
+    return new FatalError(msg);
+  };
+
+  let ReportNoPossiblyNormalCompletionError = () => {
+    let msg = "This operation is not yet supported on case blocks that result in a PossiblyNormalCompletion.";
+    realm.reportIntrospectionError(msg);
+    return new FatalError(msg);
+  };
+
+  let ReportNoAbruptCompletionError = () => {
+    let msg = "This operation is not yet supported on case blocks that result in an AbruptCompletino (by throwing).";
+    realm.reportIntrospectionError(msg);
+    return new FatalError(msg);
+  };
+
+  let DefiniteCaseEvaluation = (caseIndex: number): Value => {
+    let result = realm.intrinsics.undefined;
+    // we start at the case we've been asked to evaluate, and process statements
+    // until there is either a break statement or exception thrown (this means we
+    // implicitly fall through correctly in the absence of a break statement).
+    while (caseIndex < cases.length && !(result instanceof Completion)) {
+      let c = cases[caseIndex];
+      for (let i = 0; i < c.consequent.length && !(result instanceof Completion); ++i) {
+        let node = c.consequent[i];
+        let r = env.evaluateCompletion(node, strictCode);
+
+        // TODO understand Reference and how to deal with it better
+        // TODO come up with a strategy for handling PossiblyNormalCompletion correctly
+        if (r instanceof Reference) {
+          throw ReportNoReferenceError();
+        } else if (r instanceof PossiblyNormalCompletion) {
+          throw ReportNoPossiblyNormalCompletionError();
```

Comment:
Of course. Note that in this case I am not making up exception classes. I am calling functions with those names that return the usual FatalError. I just wanted to wrap them up rather than repeat the code to report the diagnostic and throw the error each time I wanted to do it. So is using a wrapper function like this to do what you described a problem?

Reply:
It is easy to misread this pattern, which is what happened to me. The other reason you don't want to do this is that you end up with very different looking code when you decide to actually recover from an error.

Reply:
Good point! Especially in Javascript where things are much more free flowing than my most recent implementation language (not good or bad...just different ;-)).